src/Workspaces/Core/Portable/Progress/CodeAnalysisProgress.cs

```diff
@@ -0,0 +1,149 @@
+    /// available.  This should be used sparingly.  Locations that currently do not show progress should ideally be
+    /// migrated to ones that do so that long running operations are visible to the user in a coherent fashion.
+    /// </summary>
+    internal static readonly IProgress<CodeAnalysisProgress> None = NullProgress<CodeAnalysisProgress>.Instance;
+
+    internal bool ClearValue { get; init; }
+    internal int? CompleteItemValue { get; init; }
+    internal int? IncompleteItemsValue { get; init; }
+    internal string? DescriptionValue { get; init; }
+
+    /// <summary>
+    /// Updates the UI showing the progress of the current operation to the specified <paramref name="description"/>.
+    /// </summary>
+    /// <example>
+    /// progress.Report(CodeAnalysisProgress.Description("Renaming files"));
+    /// </example>
+    public static CodeAnalysisProgress Description(string description)
+        => new() { DescriptionValue = description ?? throw new ArgumentNullException(nameof(description)) };
+
+    /// <summary>
+    /// Adds the requested number of incomplete items to the UI showing the progress of the current operation.  This is
+    /// commonly presented with a progress bar.  An optional <paramref name="description"/> can also be provided to
+    /// update the UI accordingly (see <see cref="Description"/>).
+    /// </summary>
+    /// <param name="count">The number of incomplete items left to perform.</param>
+    /// <param name="description">Optional description to update the UI to.</param>
+    /// <example>
+    /// progress.Report(CodeAnalysisProgress.AddIncompleteItems(20));
+    /// </example>
+    public static CodeAnalysisProgress AddIncompleteItems(int count, string? description = null)
+        => new()
+        {
+            IncompleteItemsValue = count >= 0 ? count : throw new ArgumentOutOfRangeException(nameof(count)),
+            DescriptionValue = description,
+        };
+
+    /// <summary>
+    /// By default, indicates that an item of work has transitioned from being incomplete (see <see
+    /// cref="AddIncompleteItems"/> to complete.  This is commonly presented with a progress bar. An optional <paramref
+    /// name="description"/> can also be provided to update the UI accordingly (see <see cref="Description"/>).
+    /// </summary>
+    /// <remarks>
+    /// Multiple items of work can be transitioned to be complete by passing an explicit value to <paramref
+    /// name="count"/>.
+    /// </remarks>
+    /// <param name="count">The number of items that were completed.  Defaults to <c>1</c> for the trivial case where
+    /// only a single item has been completed.  Must be greater than or equal to 1.</param>
+    /// <param name="description">Optional description to update the UI to.</param>
+    /// <example>
+    /// progress.Report(CodeAnalysisProgress.CompleteItem());
+    /// </example>
+    public static CodeAnalysisProgress CompleteItem(int count = 1, string? description = null)
```

Comment:
Again, probably just being nit-picky, but it seems odd that this method name is singular but can indicate multiple items completed. I understand that the default is to indicate a singular item completed, but usage like 

progress.Report(CodeAnalysisProgress.CompleteItem(10));

would get a double take from me in a code review,.

Reply:
the challenge is that hte common case is `progress.CompleteItem()`  and i wanted that to read well.  But i think i have another way to do this.

Reply:
ok.  renamed